### ISSUE TITLE
How to get a completely parameter-less functional model? 

### ISSUE THREAD

**Issue by lucaslie** [facebookresearch/higher#77]
Hi, 

thank you for this awesome library. I am excited to use it within my research. 

I have started to write some code with it and I have a question regarding the object that is returned by 'higher.patch.monkeypatch`. In particular, I would want it to never track the history of parameters it has been called with. So I want to purely use it as parameter-less function (except for tracking stats in batchnorm, etc...). 

Here is a code snippet to reproduce my question: 
```python
import copy
import higher
from torchvision.models import resnet18

net = resnet18()
fnet = higher.patch.monkeypatch(net, copy_initial_weights=False)

# get a copy of the parameters to manage myself
my_params = [param.detach().clone() for param in net.parameters()]

for _ in range(10):
    ins = torch.rand((1, 3, 224, 224)
    outs = fnet(ins, params=my_params)
    update_params(my_params)  # some arbitrary function to update my_params
```

Now if I run this code `fnet` will remember the history of parameters it has been called with. In this particular code snippet, it will have 10 copies of `my_params` stored.

Is there a good way to avoid that `fnet` is keeping track of the parameter history? 

So far I have only been able to come up with the following hack after each call to `fnet`
```python
outs = fnet(ins, params=my_params)
update_params(my_params)  # some arbitrary function to update my_params
fnet._fast_params = None
```
I have been able to avoid the memory "leak" that is otherwise created but to be honest I am not sure if I am missing something when applying this hack.

So I am wondering if there is a proper way to initialize `fnet` so I never have to worry about it in the first place? 

Thank you very much and I look forward to hearing back from you.

**Comment by lucaslie** [comment]
So I think I made some "progress" since my initial post in having a parameter-less functional model. Right now, I am initializing it like this: 
```python
fnet = higher.patch.make_functional(net)
higher.patch.buffer_sync(net, fnet)
fnet.track_higher_grads = False
fnet._fast_params = [[]]  # necessary, otherwise it throws an error during the forward pass ? 
```

Still seems quite hacky to me but at least it does the job. Do you have any suggestions on how to achieve the same thing in a better way? 

I also have a follow-up question now:

With `fnet.tracker_higher_grads = False` it seems `fnet` only keeps a reference to the latest set of parameters, i.e., it keeps a reference to `my_params` when calling `fnet(ins, params=my_params)`. Is there a way to avoid that? I am frequently calling `fnet` like this:

```python
some_parameter_modification = lambda x: 2*x
temp_params = [some_parameter_modification(param) for param in my_params]
outs = fnet(ins, params=temp_params)
```

Now I would like to avoid that `fnet` is storing a reference to `temp_params` especially in inference mode (i.e. with `torch.no_grad()`) since it's not really relevant and it will take up the memory until the reference is destroyed (and consequently the memory is freed) which won't happen unless I do another forward pass over the network. I can only think of 
```python
outs = fnet(ins, params=temp_params)
del fnet._fast_params
fnet._fast_params = [[]]
```
as some kind of solution but again that's pretty hacky. Also is there any other internal references to `temp_params` that I am missing potentially?

**Comment by lucaslie** [comment]
So I think I made some "progress" since my initial post in having a parameter-less functional model. Right now, I am initializing it like this: 
```python
fnet = higher.patch.make_functional(net)
higher.patch.buffer_sync(net, fnet)
fnet.track_higher_grads = False
fnet._fast_params = [[]]  # necessary, otherwise it throws an error during the forward pass ? 
```

Still seems quite hacky to me but at least it does the job. Do you have any suggestions on how to achieve the same thing in a better way? 

I also have a follow-up question now:

With `fnet.tracker_higher_grads = False` it seems `fnet` only keeps a reference to the latest set of parameters, i.e., it keeps a reference to `my_params` when calling `fnet(ins, params=my_params)`. Is there a way to avoid that? I am frequently calling `fnet` like this:

```python
some_parameter_modification = lambda x: 2*x
temp_params = [some_parameter_modification(param) for param in my_params]
outs = fnet(ins, params=temp_params)
```

Now I would like to avoid that `fnet` is storing a reference to `temp_params` especially in inference mode (i.e. with `torch.no_grad()`) since it's not really relevant and it will take up the memory until the reference is destroyed (and consequently the memory is freed) which won't happen unless I do another forward pass over the network. I can only think of 
```python
outs = fnet(ins, params=temp_params)
del fnet._fast_params
fnet._fast_params = [[]]
```
as some kind of solution but again that's pretty hacky. Also is there any other internal references to `temp_params` that I am missing potentially?

Thank you very much!
